pkg/controller/controller.go

```diff
@@ -1004,3 +1018,26 @@ func deprecationWarning(deprecatedParam, newParam, removalVersion string) string
 	}
 	return fmt.Sprintf("\"%s\" is deprecated and will be removed in %s%s", deprecatedParam, removalVersion, newParamPhrase)
 }
+
+func isFinalError(err error) bool {
+	// Sources:
+	// https://github.com/grpc/grpc/blob/master/doc/statuscodes.md
+	// https://github.com/container-storage-interface/spec/blob/master/spec.md
+	st, ok := status.FromError(err)
+	if !ok {
+		// This is not gRPC error. The operation must have failed before gRPC
+		// method was called, otherwise we would get gRPC error.
+		// We don't know if any previous CreateVolume is in progress, be on the safe side.
+		return false
+	}
+	switch st.Code() {
+	case codes.Canceled, // gRPC: Client Application cancelled the request
+		codes.DeadlineExceeded,  // gRPC: Timeout
+		codes.Unavailable,       // gRPC: Server shutting down, TCP connection broken - previous CreateVolume() may be still in progress.
+		codes.ResourceExhausted: // gRPC: Server temporarily out of resources - previous Provision() may be still in progress.
+		return false
+	}
+	// All other errors mean that provisioning either did not
+	// even start or failed. It is for sure not in progress.
+	return true
```

Comment:
Would it be safer to do the reverse?  Whitelist known final error codes (that we define in the csi spec), and assume all other error codes need to be retried?

Reply:
I am not sure it would bring any value. It has been working well similarly in external-attacher.

Reply:
It means that if some new error code was added or we misunderstood the meaning of some other error code, then we will do the safer operation to retry

Reply:
Said another way, by default, we will retry

Reply:
CSI spec also defines ABORTED error code as operation in progress

Reply:
Not for `CreateVolumeRequest`. IMO it's for multiple calls for a single volume and there is no volume yet.

Or is there a bug in CSI spec?

Reply:
I'm re-reading CSI spec again, it seems that ABORTED can be returned to any call... I added it to the list.